messages.go

```diff
@@ -0,0 +1,140 @@
+package dht
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/routing"
+	record "github.com/libp2p/go-libp2p-record"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+)
+
+type ProtocolMessenger struct {
+	m         *messageManager
+	validator record.Validator
+}
+
+func NewProtocolMessenger(host host.Host, protocols []protocol.ID, validator record.Validator) *ProtocolMessenger {
+	return &ProtocolMessenger{
+		m: &messageManager{
+			host:      host,
+			strmap:    make(map[peer.ID]*messageSender),
+			protocols: protocols,
+		},
+		validator: validator,
+	}
+}
+
+// putValueToPeer stores the given key/value pair at the peer 'p'
+func (pm *ProtocolMessenger) PutValue(ctx context.Context, p peer.ID, rec *recpb.Record) error {
+	pmes := pb.NewMessage(pb.Message_PUT_VALUE, rec.Key, 0)
+	pmes.Record = rec
+	rpmes, err := pm.m.sendRequest(ctx, p, pmes)
+	if err != nil {
+		logger.Debugw("failed to put value to peer", "to", p, "key", loggableKeyBytes(rec.Key), "error", err)
+		return err
+	}
+
+	if !bytes.Equal(rpmes.GetRecord().Value, pmes.GetRecord().Value) {
+		logger.Infow("value not put correctly", "put-message", pmes, "get-message", rpmes)
+		return errors.New("value not put correctly")
+	}
+
+	return nil
+}
+
+// GetValue queries a particular peer p for the value for
+// key. It returns the value and a list of closer peers.
+func (pm *ProtocolMessenger) GetValue(ctx context.Context, p peer.ID, key string) (*recpb.Record, []*peer.AddrInfo, error) {
+	pmes := pb.NewMessage(pb.Message_GET_VALUE, []byte(key), 0)
+	respMsg, err := pm.m.sendRequest(ctx, p, pmes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Perhaps we were given closer peers
+	peers := pb.PBPeersToPeerInfos(respMsg.GetCloserPeers())
+
+	if record := respMsg.GetRecord(); record != nil {
+		// Success! We were given the value
+		logger.Debug("got value")
+
+		// make sure record is valid.
+		err = pm.validator.Validate(string(record.GetKey()), record.GetValue())
+		if err != nil {
+			logger.Debug("received invalid record (discarded)")
+			// return a sentinal to signify an invalid record was received
+			err = errInvalidRecord
+			record = new(recpb.Record)
```

Comment:
I think we can get rid of this and simply return `nil, peers, errInvalidRecord` as we don't really consume this anywhere. If you look at `dht.getValues`, you'll see that this record will be ignored since `rec.GetValue()` will be nil.

Reply:
Thinking further, we don't need to special case `errInvalidRecord ` here . Let's simply log this and return the error without changing it.

Reply:
> I think we can get rid of this and simply return nil, peers, errInvalidRecord as we don't really consume this anywhere.

Sure, I think this type of defensive code tends to emerge when the "contracts" between internal components aren't super clear.

> Thinking further, we don't need to special case errInvalidRecord here . Let's simply log this and return the error without changing it.

That's actually a behavior change, which we can do later if we want (perhaps related to or after #655). The Quorum for `Get/SearchValue` currently counts invalid records towards the quorum. I think the idea here was that we would only start getting records (even invalid ones) once we were close to the target and so it was safe to give up. However, I'm not really sure that's a good idea especially now that this quorum behavior should be less necessary for the network due to the various improvements we've made (e.g. removing undialables from the routing table and having the lookup logic terminate after hearing from the Beta closest peers).

Reply:
Done